src/utils/daos_hdlr.c

```diff
@@ -2887,6 +3086,595 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+		fprintf(stderr, "failed to update destination value: %d\n", rc);
+		D_GOTO(out, rc);
+	}
+out:
+	return rc;
+}
+
+static int
+cont_clone_recx_array(daos_key_t *dkey,
+		      daos_key_t *akey,
+		      daos_handle_t *src_oh,
+		      daos_handle_t *dst_oh,
+		      daos_iod_t *iod)
+{
+	int			rc = 0;
+	int			i = 0;
+	int			k = 0;
+	uint64_t		buf_len = 0;
+	uint32_t		number = 5;
+	daos_anchor_t		recx_anchor = {0};
+	d_sg_list_t		sgl;
+	daos_epoch_range_t	eprs[5];
+	daos_recx_t		recxs[5];
+	daos_size_t		size;
+	char			**buf = NULL;
+
+	while (!daos_anchor_is_eof(&recx_anchor)) {
+		/* list all recx for this dkey/akey */
+		number = 5;
+		rc = daos_obj_list_recx(*src_oh, DAOS_TX_NONE, dkey,
+					akey, &size, &number, recxs,
+					eprs, &recx_anchor,
+					true, NULL);
+		if (rc != 0) {
+			fprintf(stderr, "failed to list recx: %d\n", rc);
+			D_GOTO(out, rc);
+		}
+
+		/* if no recx is returned for this dkey/akey move on */
+		if (number == 0)
+			continue;
+
+		D_ALLOC(buf, number);
+		if (buf == NULL) {
+			D_GOTO(out, rc = -DER_NOMEM);
+		}
+
+		d_iov_t	iov[number];
+
+		/* set iod values */
+		(*iod).iod_type  = DAOS_IOD_ARRAY;
+		(*iod).iod_nr    = number;
+		(*iod).iod_recxs = recxs;
+		(*iod).iod_size  = size;
+
+		/* set sgl values */
+		sgl.sg_nr_out = 0;
+		sgl.sg_iovs   = iov;
+		sgl.sg_nr     = number;
+
+		for (i = 0; i < number; i++) {
+			buf_len = recxs[i].rx_nr * size;
+			D_ALLOC(buf[i], buf_len);
+			if (buf[i] == NULL) {
+				/* free any other buf[i] that
+				 * was previously allocated
+				 * on error
+				 */
+				for (k = 0; k < i; k++) {
+					D_FREE(buf[k]);
+				}
+				D_FREE(buf);
+				D_GOTO(out, rc = -DER_NOMEM);
+			}
+			d_iov_set(&iov[i], buf[i], buf_len);
+		}
+
+		/* fetch recx values from source */
+		rc = daos_obj_fetch(*src_oh, DAOS_TX_NONE, 0, dkey,
+				    1, iod, &sgl, NULL, NULL);
+		if (rc != 0) {
+			fprintf(stderr, "failed to fetch source "
+				"recx: %d\n", rc);
+			D_GOTO(err_out, rc);
+		}
+
+		/* update fetched recx values and place in
+		 * destination object
+		 */
+		rc = daos_obj_update(*dst_oh, DAOS_TX_NONE, 0, dkey,
+				     1, iod, &sgl, NULL);
+		if (rc != 0) {
+			fprintf(stderr, "failed to update destination "
+				"recx: %d\n", rc);
+			D_GOTO(err_out, rc);
+		}
+		for (i = 0; i < number; i++) {
+			D_FREE(buf[i]);
+		}
+		D_FREE(buf);
+	}
+	D_GOTO(out, rc);
+err_out:
+	for (i = 0; i < number; i++) {
+		D_FREE(buf[i]);
+	}
+	D_FREE(buf);
+out:
+	return rc;
+}
+
+static int
+cont_clone_list_akeys(daos_handle_t *src_oh,
+		      daos_handle_t *dst_oh,
+		     daos_key_t diov)
+{
+	int		rc = 0;
+	int		j = 0;
+	char		*ptr;
+	daos_anchor_t	akey_anchor = {0};
+	daos_key_desc_t	akey_kds[ENUM_DESC_NR] = {0};
+	uint32_t	akey_number = ENUM_DESC_NR;
+	char		*akey = NULL;
+	d_sg_list_t	sgl;
+	d_iov_t		iov;
+	daos_key_t	aiov;
+	daos_iod_t	iod;
+	char		*small_key = NULL;
+	char		*large_key = NULL;
+	char		*key_buf = NULL;
+	daos_size_t	key_buf_len = 0;
+
+	D_ALLOC(small_key, ENUM_DESC_BUF);
```

Comment:
future improvement:
for small keys buffer, it's probably better to just have that on the stack to avoid too many allocations.

Reply:
@mchaarawi Something I've thought about is the possibility of passing buffers _**to**_ functions like this - maybe in a struct. Then, we could start with a small alloc that is done a single time and passed with each call. In this function, we could realloc as needed for larger buffers.
Of course, there are some caveats, but what do you think about this general idea?